### DISCUSSION TITLE
Road map

### DISCUSSION THREAD

**Discussion by DevNullx64** [Leonardo16AM/Sharp-Grad#10] [General]
What is you feeling about the next milles stone ?

**Comment by Leonardo16AM** [comment]
Hi, I think it will depend on what we want to achieve. What do you want to achieve? We can create a roadmap being guided by your wanted goals.

I have some ideas:

Without using the GPU:
 - Create a Tensor Class
 - Add tensorial operations (Dot Productm Cross product etc)
 -  Build the NN Module over Tensorial Operations
 - 
 To use the GPU we should investigate the best way to do it, maybe watching how does  Torch and Tensorflow works will be helpfull.

**Reply by DevNullx64** [reply]
It's look like greate !
Do you have Iedea for tensor Class ? Most blocking point for me is Shape and Reshape.
If we can, keep a zero memory copie reshape it would be greate.
Shpe seem to have some restriction based on the computation too. a shape of (m, k) matmul (k, n) result in a shape of (m, n). I'm not familiar whit that

**Reply by Leonardo16AM** [reply]
I watched the way in which you coded the Shape, and I think it is very optimal, the tensor is not shape-dependant, because all the data are stores in a linear array and you calculate the positions of the elements using GetFlattenedIndex() . So, reshaping it  will not be only changing Tensor.Shape()?  What do you mean by zero memory?   

With tensors we different types of multiplications, Tensor Multiplication, Matrix Multiplication and Pairwise Multiplication (and Dot Product):
In Tensor Multiplication you create basically a new tensor with the shape being basically the join of the two shapes, for example: 

$$
\[
A = \begin{pmatrix}
a_{11} & a_{12} \\
a_{21} & a_{22}
\end{pmatrix}, \quad
B = \begin{pmatrix}
b_{11} & b_{12} \\
b_{21} & b_{22}
\end{pmatrix}
\]
$$

The tensorial product is: 

$$
\[
A \otimes B = \begin{pmatrix}
a_{11} \begin{pmatrix}
b_{11} & b_{12} \\
b_{21} & b_{22}
\end{pmatrix} & a_{12} \begin{pmatrix}
b_{11} & b_{12} \\
b_{21} & b_{22}
\end{pmatrix} \\
a_{21} \begin{pmatrix}
b_{11} & b_{12} \\
b_{21} & b_{22}
\end{pmatrix} & a_{22} \begin{pmatrix}
b_{11} & b_{12} \\
b_{21} & b_{22}
\end{pmatrix}
\end{pmatrix}
\]
$$
$$
\[
A \otimes B = \begin{pmatrix}
a_{11}b_{11} & a_{11}b_{12} & a_{12}b_{11} & a_{12}b_{12} \\
a_{11}b_{21} & a_{11}b_{22} & a_{12}b_{21} & a_{12}b_{22} \\
a_{21}b_{11} & a_{21}b_{12} & a_{22}b_{11} & a_{22}b_{12} \\
a_{21}b_{21} & a_{21}b_{22} & a_{22}b_{21} & a_{22}b_{22}
\end{pmatrix}
\]
$$

On the other side, the Matrix multiplication of two matrices of shape M*K and K*N creates a new matrix of size M*N, Matrix multiplication is actually the composition of two operations: outer product and tensor contraction  and it works like in this way:

``` dotnet
for (int i = 0; i < M; i++)
{
      for (int j = 0; j < N; j++)
      {
                C[i, j] = 0;
                for (int k = 0; k < K; k++)
                {
                           C[i, j] += A[i, k] * B[k, j];
                }
      }
}
```

**Reply by DevNullx64** [reply]
I worry because of I maid lot of assertion and I can't confirm all myself. I try to search help with collegue but in real I explain the most...
So now I can confirm that a tensor can be handled as a linear array. With coordinate computation bound by the Shape.
In this case, yes, reshare is a zero memory copy operation. Array itself is lefts as is, juste the Vector struct should be insatnciate with new parameters. I'm really appy about this informations :) Vector struct use some octet (2 array pointer + shape coordinate). It's "costless" :)

With what you said I think that I can handle the first, and real, tensor multiplication. For the two other we need operator as fonction. Same as Pow in C# because the '^' symbole (mostly used for exponetiation) is already used for XOR operation. But I thnik that add the possibility to the shape to be computable base on tensor opération is something to handle.
I Will try my best to hide all the complexity and whish that add a new opérator is just fill a new class or struct.

**Reply by Leonardo16AM** [reply]
Maybe it will be good to name the operations (and operators) equal to pytorch , just for familiarity. 
They use:
 -  a*b ormul(a,b) for element-wise multiplication
 -  a@b or matmult(a,b) for matrix multiplication
 I don't remember the other one, there is also dot(a,b) tensordot(a,b,dims), kron(a,b)   (Kronecker product) etc.

**Comment by DevNullx64** [comment]
FMI: Do you want to become a PhD graduate and publish as a researcher, or do you plan to study and then enter the private sector?

**Reply by Leonardo16AM** [reply]
Probably after finishing my Bachelor degree I will try to do a PHD and then enter to the private sector, or maybe work directly in the private sector for a few years and then try to do a PHD
